### PR TITLE
improve escaped-quote handling in sys.process (CommandLineParser)

### LINKED ISSUE #12611: sys.process: parser for shell tokenization mishandles some quotes+escapes
In the sys.process context, this is a 2.13 regression; the code in question was added on 2.13.x in scala/scala#7452 (2018)

(Though the copy of the code in `CommandLineParser` has had the bug for longer; we should fix it both places.)

customer writes:

> The bug is that the parser for shell tokenization doesn't check for escape characters the first time it encounters a quote

```scala
scala 2.13.8> import sys.process._
import sys.process._

scala 2.13.8> """echo '\"hello world!\"' """.!!
val res1: String =
"\"hello world!\"
"
```

so far so good, but then:

```scala
scala 2.13.8> """echo \"hello world!\" """.!!
scala.sys.process.Parser$ParseException: Unmatched quote
  at scala.sys.process.Parser$.tokenize(Parser.scala:111)
  at scala.sys.process.ProcessCreation.apply(Process.scala:95)
```

which worked in 2.12.16.

### PR DESCRIPTION
**Pull request by SethTisue**
fixes scala/bug#12611

### CODE REVIEW

File: src/compiler/scala/tools/cmd/CommandLineParser.scala [hunk 1]
```diff
@@ -14,6 +14,11 @@ package scala.tools.cmd
 
 import scala.annotation.tailrec
 
+// This and scala/sys/process/Parser.scala are copies of each other
+// and should be maintained in parallel. (Ideally we'd eliminate
+// the duplication, which arose because we expected sys.process to
+// become a separate module.)
+
```
Comment (som-snytt) [inline src/compiler/scala/tools/cmd/CommandLineParser.scala]: `object CommandLineParser extends scala.sys.process.Parser`

File: src/library/scala/sys/process/Parser.scala [hunk 2]
```diff
@@ -14,6 +14,11 @@ package scala.sys.process
...
 import scala.annotation.tailrec
 
+// This and scala/tools/cmd/CommandLineParser.scala are copies of each other
+// and should be maintained in parallel. (Ideally we'd eliminate
+// the duplication, which arose because we expected sys.process to
+// become a separate module.)
+
 /** A simple enough command line parser.
  */
 private[scala] object Parser {
```
Comment (som-snytt) [inline src/library/scala/sys/process/Parser.scala:19]:
```
-private[scala] object Parser {
+private[scala] class Parser {

-      def terminal = in.charAt(pos) match {
+      def terminal: Boolean = cur match {

-    @tailrec
-    def skipToDelim(): Boolean =
-      cur match {
+    // Skip to a word boundary, where words can be quoted and quotes can be escaped
+    def skipToDelim(): Boolean = {
+      var escaped = false
+      def advance(): Boolean = cur match {
+        case _ if escaped         => escaped = false ; bump() ; advance()
+        case '\\'                 => escaped = true ; bump() ; advance()
         case q @ (DQ | SQ)        => { qpos += pos; bump(); skipToQuote(q) } && { qpos += pos; bump(); skipToDelim() }
         case -1                   => true
         case c if isWhitespace(c) => true
         case _                    => bump(); skipToDelim()
       }
+      advance()
+    }

+object Parser extends Parser
```
I could push to branch if the tweaks look good.

File: src/library/scala/sys/process/Parser.scala [hunk 3]
```diff
@@ -49,9 +54,12 @@ private[scala] object Parser {
       while (!done && !terminal) pos += 1
       !done
     }
+    var escapedCur = false
```
Comment (som-snytt) [inline src/library/scala/sys/process/Parser.scala]: My suggestion makes it more local.

**Review by som-snytt — APPROVED** [review]
So much for "at least I'll never have to look at that code ever again."

I could push tweaks to branch or follow up when I forward port it.

**Comment by SethTisue** [comment]
@som-snytt sure, you can push your suggestions to my branch

**Comment by som-snytt** [comment]
Without waiting for the slow green check, I will also push Seth's suggestion to bite the bullet and remove the `cmd` package. There are just a handful of usages.

**Comment by som-snytt** [comment]
Squashed my commits, including removal of `tools.cmd`, because of funky mima -- it complained about missing ctors for `ParseException`, when `class Parser {...}; object Parser extends Parser`. That form was unnecessary after removing `CommandLineParser extends Parser`.

Fixed tests which "leveraged" `CommandLineParser`.

Edit:
```
[error] /home/jenkins/workspace/scala-2.13.x-validate-main/test/benchmarks/src/main/scala/scala/tools/cmd/CommandLineParserBenchmark.scala:18:10: not found: value CommandLineParser
[error] 1 of 19 test tasks failed:
[error] - bench/compile
[error]   - bench/compile:compileIncremental failed: Compilation failed
[error] Failure due to previous errors
[error] (testAll) Failure due to previous errors
```

**Comment by som-snytt** [comment]
Requires Seth's review of following up his suggestions on the ticket and thereby creating more work for himself.

**Pull request merged**